### PR TITLE
registry: correctly bump tag expiration when tag changes upstream (PROJQUAY-3976)

### PR DESCRIPTION
**Pull request by flavianmissi**
fixes [PROJQUAY-3976](https://issues.redhat.com//browse/PROJQUAY-3976)

### CODE REVIEW

File: data/registry_model/registry_proxy_model.py [hunk 1]
```diff
@@ -185,13 +191,36 @@ def lookup_manifest_by_digest(
...
+                    .order_by_extend(TagTable.lifetime_end_ms.desc())
                 )
-                parent_ids = [link.manifest_id for link in links]
+                items = (
+                    Select(columns=[subq.c.rnk, subq.c.id, subq.c.name, subq.c.lifetime_end_ms])
+                    .from_(subq)
+                    .where(subq.c.rnk == 1)
+                    .bind(db)
+                )
+                tag_ids = [item["id"] for item in items]
```
Comment (flavianmissi) [inline data/registry_model/registry_proxy_model.py:220]: this query turned out way more complicated than I wanted it to be. I started using `DISTINCT ON` but sqlite doesn't support that (only `DISTINCT` without `ON`), so this is how I managed to make the same query work on both postgres and sqlite.

Comment (flavianmissi) [reply data/registry_model/registry_proxy_model.py:220]: ugh typical
https://github.com/quay/quay/runs/7790348396?check_suite_focus=true#step:5:1181

File: data/registry_model/registry_proxy_model.py [hunk 2]
```diff
@@ -185,13 +191,36 @@ def lookup_manifest_by_digest(
...
+                    .join(ManifestChild, on=(TagTable.manifest_id == ManifestChild.manifest_id))
+                    .where(
+                        (ManifestChild.repository_id == repository_ref.id)
+                        & (ManifestChild.child_manifest_id == wrapped_manifest.id)
+                    )
+                    .order_by(TagTable.name)
+                    .order_by_extend(TagTable.lifetime_end_ms.desc())
                 )
-                parent_ids = [link.manifest_id for link in links]
+                items = (
```
Comment (dmage) [inline data/registry_model/registry_proxy_model.py:214]: Will this query use indexes?

Comment (flavianmissi) [reply data/registry_model/registry_proxy_model.py:214]: I'm not sure about the order by name tbh, but that's there for troubleshooting purposes so I'll remove it. the rest should be using the indexes.

**Review by HammerMeetNail — APPROVED** [review]
LGTM 👍

**Review by HammerMeetNail — APPROVED** [review]

**Pull request merged**